### PR TITLE
North Dakota state income tax

### PR DESCRIPTION
**Pull request by martinholmer**
This pull request adds the North Dakota state income tax.

The code in this pull request generates the same North Dakota state income tax as the patched version of TAXSIM35 described in issue #993 for the following samples of 100,000 randomly-generated integration test cases: `p21`, `e21`, `f21`, `g21`, `h21`,  `i21`, and `j21`.  Out of the 100,000 `k21` sample test cases, there are 8834 differences, all of which are caused by the same problem: the current version of PolicyEngineUS does not have a way of calculating correct state or federal income taxes when circular logic errors prevent the state income variable from being included in the `state_income_tax` variable's list of state income taxes and the test case has itemized expenses of certain amounts.

The circular logic problem arises for North Dakota because ND taxable income is defined as US taxable income plus some additions minus some subtractions.  So, if a tax unit has itemized expenses (other than the ND state income tax), there is a simultaneous equation situation between the US and ND tax.  The current version of PolicyEngineUS avoids the circular logic errors by assuming in the US income tax calculations that the ND income tax amount is zero.  In some cases this assumption leads to an overestimated US taxable income amount (because of the underestimated itemizable expenses), and hence, to overestimated ND taxable income and income tax amounts.  A variant of this problem has been discussed in issue #1827.

### CODE REVIEW

File: policyengine_us/parameters/gov/states/nd/tax/income/taxable_income/additions/sources.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+description: ND additions to US taxable income.
+values:
+  2021-01-01:
+    - c05700  # lump-sum distributions from Federal Form 4972
+
+metadata:
+  unit: variable
+  label: ND additions to US taxable income
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/nd/tax/income/taxable_income/additions/sources.yaml:8]:
```suggestion
  label: North Dakota additions
```

File: policyengine_us/parameters/gov/states/nd/tax/income/taxable_income/additions/sources.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+description: ND additions to US taxable income.
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/nd/tax/income/taxable_income/additions/sources.yaml:1]:
```suggestion
description: North Dakota adds these variables to US taxable income when computing its taxable income.
```

File: policyengine_us/parameters/gov/states/nd/tax/income/taxable_income/additions/sources.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    - c05700  # lump-sum distributions from Federal Form 4972
+
+metadata:
+  unit: variable
+  label: ND additions to US taxable income
+  reference:
+  - title: 2021 ND income tax form
+    href: https://www.tax.nd.gov/sites/www/files/documents/forms/form-nd-1-2021.pdf
+  - title: 2021 ND income tax instructions
+    href: https://www.tax.nd.gov/sites/www/files/documents/forms/2021-individual-income-tax-booklet.pdf
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/nd/tax/income/taxable_income/additions/sources.yaml:13]: If not too much trouble could you add the page number? `#page=x`

File: policyengine_us/variables/gov/states/nd/tax/income/nd_taxable_income.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from policyengine_us.model_api import *
+
+
+class nd_taxable_income(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "ND taxable income"
+    unit = currency-USD
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/nd/tax/income/nd_taxable_income.py:8]: (Sorry, it differs from the parameter convention)
```suggestion
    unit = USD
```

**Review by MaxGhenis — COMMENTED** [review]
Great to see this. Just dropping some of our newer style conventions if you could please apply them.

File: policyengine_us/parameters/gov/states/nd/tax/income/taxable_income/additions/sources.yaml (hunk 1 above)
Comment (martinholmer) [reply policyengine_us/parameters/gov/states/nd/tax/income/taxable_income/additions/sources.yaml:8]: Done

File: policyengine_us/parameters/gov/states/nd/tax/income/taxable_income/additions/sources.yaml (hunk 3 above)
Comment (martinholmer) [reply policyengine_us/parameters/gov/states/nd/tax/income/taxable_income/additions/sources.yaml:13]: It is more than a little trouble, but it is done.

File: policyengine_us/parameters/gov/states/nd/tax/income/taxable_income/additions/sources.yaml (hunk 2 above)
Comment (martinholmer) [reply policyengine_us/parameters/gov/states/nd/tax/income/taxable_income/additions/sources.yaml:1]: Done

File: policyengine_us/variables/gov/states/tax/income/state_income_tax.py [hunk 5]
```diff
@@ -15,6 +15,7 @@ class state_income_tax(Variable):
         "ma_income_tax",
         "md_income_tax",
         # "mo_income_tax",  --- activating will cause circular logic errors
+        # "nd_income_tax",  --- activating will cause circular logic errors
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/tax/income/state_income_tax.py:18]: Where's the circular reference?

**Review by MaxGhenis — COMMENTED** [review]
Thanks Martin - we can make some nonfunctional changes like labels and legislative references in subsequent PRs.

Could you please do these last tasks so it flows through to the app?
* Add to `state_income_tax_before_refundable_credits`, `household_tax_before_refundable_credits`, and `household_refundable_tax_credits`
* Add `parameters/gov/states/nd/README.md` with `# North Dakota`
* Add to `modelled_policies.yaml`

File: policyengine_us/variables/gov/states/tax/income/state_income_tax.py (hunk 5 above)
Comment (martinholmer) [reply policyengine_us/variables/gov/states/tax/income/state_income_tax.py:18]: Described in the first comment in [this pull request](#2042) discussion.

**Comment by martinholmer** [comment]
@MaxGhenis said:
> Could you please do these last tasks so it flows through to the app?
> - Add to `state_income_tax_before_refundable_credits`, `household_tax_before_refundable_credits`, and `household_refundable_tax_credits`
> - Add `parameters/gov/states/nd/README.md` with # North Dakota
> - Add to `modelled_policies.yaml`

OK, but where are all these requirements in the contributor guide?  If they aren't there, you need to add them so that contributors know what is expected of them.

**Comment by MaxGhenis** [comment]
Thanks @martinholmer, filed https://github.com/PolicyEngine/policyengine-us/issues/2065.

**Pull request merged**